cmd/apiProviderUpload_test.go

```diff
@@ -0,0 +1,121 @@
+package cmd
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"testing"
+
+	"github.com/SAP/jenkins-library/pkg/apim"
+	piperhttp "github.com/SAP/jenkins-library/pkg/http"
+	"github.com/SAP/jenkins-library/pkg/mock"
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestRunApiProviderUpload(t *testing.T) {
+	t.Parallel()
+
+	t.Run("happy path", func(t *testing.T) {
+		file, tmpErr := ioutil.TempFile("", "test.json")
+		if tmpErr != nil {
+			t.FailNow()
+		}
+		defer os.RemoveAll(file.Name()) // clean up
+		filesMock := mock.FilesMock{}
+		filesMock.AddFile(file.Name(), []byte("Test content"))
+		config := getDefaultOptionsForApiProvider()
+		config.FilePath = file.Name()
+		httpClientMock := httpMockAPIM{StatusCode: 201, ResponseBody: ``}
+		apim := apim.APIMBundle{APIServiceKey: config.APIServiceKey, Client: &httpClientMock}
+		// test
+		err := createApiProvider(&config, apim, filesMock.FileRead)
+		// assert
+		if assert.NoError(t, err) {
+			t.Run("check file", func(t *testing.T) {
+				assert.Equal(t, fileExists(file.Name()), true)
+			})
+
+			t.Run("check url", func(t *testing.T) {
+				assert.Equal(t, "/apiportal/api/1.0/Management.svc/APIProviders", httpClientMock.URL)
+			})
+			t.Run("check method", func(t *testing.T) {
+				assert.Equal(t, "POST", httpClientMock.Method)
+			})
+		}
+	})
+
+	t.Run("Failure Path", func(t *testing.T) {
+		file, tmpErr := ioutil.TempFile("", "test.json")
+		if tmpErr != nil {
+			t.FailNow()
+		}
+		defer os.RemoveAll(file.Name()) // clean up
+		filesMock := mock.FilesMock{}
+		filesMock.AddFile(file.Name(), []byte("Test content"))
+		config := getDefaultOptionsForApiProvider()
+		config.FilePath = file.Name()
+		httpClientMock := httpMockAPIM{StatusCode: 400, ResponseBody: ``}
+		apim := apim.APIMBundle{APIServiceKey: config.APIServiceKey, Client: &httpClientMock}
+		// test
+		err := createApiProvider(&config, apim, filesMock.FileRead)
+		// assert
+		assert.Error(t, err)
+	})
+
+}
+
+func getDefaultOptionsForApiProvider() apiProviderUploadOptions {
+	return apiProviderUploadOptions{
+		APIServiceKey: `{
+			"oauth": {
+				"url": "https://demo",
+				"clientid": "demouser",
+				"clientsecret": "******",
+				"tokenurl": "https://demo/oauth/token"
+			}
+		}`,
+		FilePath: "test.json",
+	}
+}
+
+type httpMockAPIM struct {
```

Comment:
All the mock functions would be better found in the apim package, as they most likely will be used by all apim commands. My fear is that the other apim commands will consume it from here, as was done in cpi.

Reply:
this is the case with almost all commands which uses http mock for their usage

Reply:
So you are saying that other http-call-commands all have their mock inside the test function of the command? And because all other http-call-commands have the mock in the command test, you do not want to move it to the pkg?
In that case: I disagree. Just because the other commands do it like that, I don't see the disadvantage of having the mock in the pkg. The opposite is the case I find. Is there another reason not to move it?

Reply:
i was meant to say, this function returns object apiProviderUploadOptions which is very specific to this command. so i cant move it to pkg.

Reply:
Don't move the function above the mock, called `getDefaultOptionsForApiProvider`. I'm saying to move the mock, specifically:
- `httpMockAPIM struct`
- ` func (c *httpMockAPIM) SetOptions(options piperhttp.ClientOptions)`
- ` func (c *httpMockAPIM) SendRequest(method string, url string, r io.Reader, header http.Header, cookies []*http.Cookie) (*http.Response, error)`

I don't see that as specific to this step. Or am I missing something?

Reply:
Spoke to @radsoulbeard about this. He says that it is fine in this file for now. When it is used by another file latest, he also suggests you move it to the pkg. 
My opinion is that the other Piper colleagues that have done it like this should not be our guides here. If we can make our tests better than they are, let's do it.